src/riak_kv_sweeper_state.erl

```diff
@@ -0,0 +1,1076 @@
+                {{0,0}, _} -> [];
+                {_N, Sweep} -> Sweep
+            end
+    end.
+
+-spec expired_or_missing(erlang:timestamp(),
+                         #sweep{},
+                         dict()) ->
+    {MissingSum :: non_neg_integer(),
+     ExpiredSum :: non_neg_integer()}.
+expired_or_missing(Now, #sweep{results = Results}, Participants) ->
+    ResultsList = dict:to_list(Results),
+    Missing = missing(run_interval, Participants, ResultsList),
+    Expired =
+        [begin
+             RunInterval = run_interval(Mod, Participants),
+             expired(Now, TS, RunInterval)
+         end ||
+         {Mod, {TS, _Outcome}} <- ResultsList],
+    MissingSum = lists:sum(Missing),
+    ExpiredSum = lists:sum(Expired),
+    {MissingSum, ExpiredSum}.
+
+-spec expired(erlang:timestamp(),
+              erlang:timestamp(),
+              RunInterval :: non_neg_integer() | disabled) ->
+    non_neg_integer().
+expired(_Now, _TS, disabled) ->
+    0;
+expired(Now, TS, RunInterval) ->
+    case elapsed_secs(Now, TS) - RunInterval of
+        N when N < 0 ->
+            0;
+        N ->
+            N
+    end.
+
+-spec run_interval(riak_kv_sweeper:participant_module(),
+                   Participants :: dict()) ->
+    RunInterval :: non_neg_integer() | 'disabled'.
+run_interval(Mod, Participants) ->
+    case dict:find(Mod, Participants) of
+        {ok, #sweep_participant{run_interval = RunInterval}} ->
+            get_run_interval(RunInterval);
+        _ ->
+            %% Participant have been disabled since last run.
+            %% TODO: should we remove inactive results?
+            disabled
+    end.
+
+-spec elapsed_secs(erlang:timestamp(), erlang:timestamp()) ->
+    non_neg_integer().
+elapsed_secs(Now, Start) ->
+    timer:now_diff(Now, Start) div 1000000.
+
+-spec random_sweep([#sweep{}]) -> #sweep{}.
+random_sweep(Sweeps) ->
+    Index = random:uniform(length(Sweeps)),
+    lists:nth(Index, Sweeps).
+
+-spec scheduler_enabled() -> boolean().
+scheduler_enabled() ->
+    case app_helper:get_env(riak_kv, sweeper_scheduler, true) of
+        E when E == true; E == false ->
+            E;
+        _ ->
+            false
+    end.
+
+-spec sweep_window() ->
+    'always'
+    | 'never'
+    | {StartHour :: non_neg_integer(), EndHour :: non_neg_integer()}.
+sweep_window() ->
+    case application:get_env(riak_kv, sweep_window) of
+        {ok, always} ->
+            always;
+        {ok, never} ->
+            never;
+        {ok, {StartHour, EndHour}} when StartHour >= 0, StartHour =< 23,
+                                        EndHour >= 0, EndHour =< 23 ->
+            {StartHour, EndHour};
+        Other ->
+            lager:warning("Invalid riak_kv_sweep window specified: ~p. "
+                          "Defaulting to 'never'.", [Other]),
+            never
+    end.
+
+-spec get_concurrency_limit() -> non_neg_integer().
+get_concurrency_limit() ->
+    case app_helper:get_env(riak_kv, sweep_concurrency,
```

Comment:
Normally Cuttlefish should guard against non-integer values here, so it's not a huge risk, but it is possible a user might go screwing around with advanced.config and put a bad value in there. Better safe than sorry, I suppose.

Now that I look at this function, I'm also not sure why we have that `case` statement there at all, since `?DEFAULT_SWEEP_CONCURRENCY` is already an integer. Will fix that as well.